File: ai-dungeon-master-offline/app/sidecar/python/memory/repositories.py
import json
import uuid
import re
from typing import List, Dict, Any, Optional
from memory.db_manager import DatabaseManager
import datetime

class CampaignRepository:
    def __init__(self, db: DatabaseManager):
        self.db = db
        self._ensure_columns()

    def _ensure_columns(self):
        conn = self.db.get_connection()
        try:
            conn.execute("ALTER TABLE campaigns ADD COLUMN narrative_summary TEXT;")
            conn.commit()
        except Exception:
            pass
        conn.close()

    def create(self, name: str, system: str, tone: str) -> Dict[str, Any]:
        campaign_id = str(uuid.uuid4())
        query = """
            INSERT INTO campaigns (id, name, system, tone, narrative_summary)
            VALUES (?, ?, ?, ?, '')
        """
        self.db.execute(query, (campaign_id, name, system, tone))
        return self.get_by_id(campaign_id)

    def get_by_id(self, campaign_id: str) -> Optional[Dict[str, Any]]:
        query = "SELECT * FROM campaigns WHERE id = ?"
        return self.db.fetch_one(query, (campaign_id,))

    def list_all(self) -> List[Dict[str, Any]]:
        query = "SELECT * FROM campaigns ORDER BY created_at DESC"
        return self.db.fetch_all(query)

    def update_summary(self, campaign_id: str, summary: str):
        query = "UPDATE campaigns SET narrative_summary = ? WHERE id = ?"
        self.db.execute(query, (summary, campaign_id))


class CharacterRepository:
    def __init__(self, db: DatabaseManager):
        self.db = db
        self._ensure_nullable_campaign()

    def _ensure_nullable_campaign(self):
        conn = self.db.get_connection()
        try:
            cursor = conn.cursor()
            cursor.execute("PRAGMA table_info(characters);")
            columns = cursor.fetchall()
            campaign_id_col = next((c for c in columns if c["name"] == "campaign_id"), None)
            
            if campaign_id_col and campaign_id_col["notnull"] == 1:
                print("[Migration] Recreating characters table to make campaign_id nullable...")
                conn.execute("PRAGMA foreign_keys=OFF;")
                conn.execute("BEGIN TRANSACTION;")
                conn.execute("ALTER TABLE characters RENAME TO characters_old;")
                conn.execute("""
                    CREATE TABLE IF NOT EXISTS characters (
                        id TEXT PRIMARY KEY,
                        campaign_id TEXT,
                        name TEXT NOT NULL,
                        class TEXT NOT NULL,
                        race TEXT NOT NULL,
                        background TEXT NOT NULL,
                        level INTEGER NOT NULL DEFAULT 1,
                        hp_current INTEGER NOT NULL,
                        hp_max INTEGER NOT NULL,
                        armor_class INTEGER NOT NULL,
                        stats_json TEXT NOT NULL,
                        inventory_json TEXT NOT NULL,
                        xp INTEGER NOT NULL DEFAULT 0,
                        FOREIGN KEY(campaign_id) REFERENCES campaigns(id) ON DELETE CASCADE
                    );
                """)
                # Copy existing data
                xp_select = "xp" if any(c["name"] == "xp" for c in columns) else "0"
                conn.execute(f"""
                    INSERT INTO characters (id, campaign_id, name, class, race, background, level, hp_current, hp_max, armor_class, stats_json, inventory_json, xp)
                    SELECT id, campaign_id, name, class, race, background, level, hp_current, hp_max, armor_class, stats_json, inventory_json, {xp_select} FROM characters_old;
                """)
                conn.execute("DROP TABLE characters_old;")
                conn.execute("COMMIT;")
                conn.execute("PRAGMA foreign_keys=ON;")
                print("[Migration] Recreated characters table successfully.")
        except Exception as e:
            print(f"[Migration] Failed to migrate characters table: {e}")
            try:
                conn.execute("ROLLBACK;")
            except Exception:
                pass
        finally:
            conn.close()

    def create(self, campaign_id: Optional[str], name: str, char_class: str, race: str, background: str, hp_max: int, armor_class: int, stats: dict, inventory: dict) -> Dict[str, Any]:
        char_id = str(uuid.uuid4())
        query = """
            INSERT INTO characters (id, campaign_id, name, class, race, background, hp_current, hp_max, armor_class, stats_json, inventory_json)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """
        self.db.execute(query, (
            char_id, campaign_id, name, char_class, race, background, hp_max, hp_max, armor_class,
            json.dumps(stats), json.dumps(inventory)
        ))
        return self.get_by_id(char_id)

    def get_by_id(self, char_id: str) -> Optional[Dict[str, Any]]:
        char = self.db.fetch_one("SELECT * FROM characters WHERE id = ?", (char_id,))
        if char:
            char = dict(char)
            char["stats"] = json.loads(char["stats_json"])
            char["inventory"] = json.loads(char["inventory_json"])
        return char

    def get_by_campaign(self, campaign_id: str) -> List[Dict[str, Any]]:
        chars = self.db.fetch_all("SELECT * FROM characters WHERE campaign_id = ?", (campaign_id,))
        result = []
        for char in chars:
            c = dict(char)
            c["stats"] = json.loads(c["stats_json"])
            c["inventory"] = json.loads(c["inventory_json"])
            result.append(c)
        return result

    def get_all_characters(self) -> List[Dict[str, Any]]:
        chars = self.db.fetch_all("SELECT * FROM characters")
        result = []
        for char in chars:
            c = dict(char)
            c["stats"] = json.loads(c["stats_json"])
            c["inventory"] = json.loads(c["inventory_json"])
            result.append(c)
        return result

    def get_global_characters(self) -> List[Dict[str, Any]]:
        chars = self.db.fetch_all("SELECT * FROM characters WHERE campaign_id IS NULL")
        result = []
        for char in chars:
            c = dict(char)
            c["stats"] = json.loads(c["stats_json"])
            c["inventory"] = json.loads(c["inventory_json"])
            result.append(c)
        return result

    def assign_to_campaign(self, char_id: str, campaign_id: str) -> bool:
        query = "UPDATE characters SET campaign_id = ? WHERE id = ?"
        return self.db.execute(query, (campaign_id, char_id)) > 0

    def update_hp(self, char_id: str, hp: int) -> bool:
        query = "UPDATE characters SET hp_current = ? WHERE id = ?"
        return self.db.execute(query, (hp, char_id)) > 0

    def update_xp(self, char_id: str, xp_gain: int) -> Dict[str, Any]:
        """Increases XP and handles level up (simplified D&D 5e thresholds)."""
        char = self.get_by_id(char_id)
        if not char:
            return {}
        xp_thresholds = [0, 300, 900, 2700, 6500, 14000, 23000, 34000, 48000, 64000, 85000, 100000, 120000, 140000, 165000, 195000, 225000, 265000, 305000, 355000]
        current_xp = char.get("xp", 0) + xp_gain
        current_level = char.get("level", 1)
        new_level = current_level
        for i, threshold in enumerate(xp_thresholds):
            if current_xp >= threshold:
                new_level = i + 1
        query = "UPDATE characters SET xp = ?, level = ? WHERE id = ?"
        self.db.execute(query, (current_xp, new_level, char_id))
        return {"xp": current_xp, "level": new_level, "leveled_up": new_level > current_level}


class DiceRollRepository:
    def __init__(self, db: DatabaseManager):
        self.db = db

    def save_roll(self, campaign_id: str, actor_id: str, actor_name: str, formula: str, result: dict) -> Dict[str, Any]:
        roll_id = str(uuid.uuid4())
        query = """
            INSERT INTO dice_rolls (id, campaign_id, actor_id, actor_name, formula, result_json)
            VALUES (?, ?, ?, ?, ?, ?)
        """
        self.db.execute(query, (roll_id, campaign_id, actor_id, actor_name, formula, json.dumps(result)))
        return {"id": roll_id, "campaign_id": campaign_id, "actor_id": actor_id, "actor_name": actor_name, "formula": formula, "result": result}


class MessageRepository:
    """Persists campaign message history for LLM context window with book page details."""

    def __init__(self, db: DatabaseManager):
        self.db = db
        self._ensure_table()

    def _ensure_table(self):
        conn = self.db.get_connection()
        conn.execute("""
            CREATE TABLE IF NOT EXISTS campaign_messages (
                id TEXT PRIMARY KEY,
                campaign_id TEXT NOT NULL,
                character_id TEXT,
                role TEXT NOT NULL,
                content TEXT NOT NULL,
                page_number INTEGER DEFAULT 1,
                coordinates_json TEXT,
                choices_json TEXT,
                mechanics TEXT,
                is_active INTEGER DEFAULT 1,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            );
        """)
        # Alter table to add columns if they don't exist (compatibility layer)
        columns = [
            ("page_number", "INTEGER DEFAULT 1"),
            ("coordinates_json", "TEXT"),
            ("choices_json", "TEXT"),
            ("mechanics", "TEXT"),
            ("is_active", "INTEGER DEFAULT 1")
        ]
        for col_name, col_type in columns:
            try:
                conn.execute(f"ALTER TABLE campaign_messages ADD COLUMN {col_name} {col_type};")
            except Exception:
                pass # Column already exists
        try:
            conn.execute("CREATE INDEX IF NOT EXISTS idx_messages_campaign ON campaign_messages (campaign_id, is_active, page_number);")
        except Exception:
            pass
        conn.commit()
        conn.close()

    def add_message(self, campaign_id: str, role: str, content: str, character_id: Optional[str] = None, page_number: int = 1, coordinates_json: Optional[str] = None, choices_json: Optional[str] = None, mechanics: Optional[str] = None) -> None:
        msg_id = str(uuid.uuid4())
        self.db.execute(
            """INSERT INTO campaign_messages (id, campaign_id, character_id, role, content, page_number, coordinates_json, choices_json, mechanics, is_active)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 1)""",
            (msg_id, campaign_id, character_id, role, content, page_number, coordinates_json, choices_json, mechanics)
        )

    def get_recent_history(self, campaign_id: str, limit: int = 10) -> List[Dict[str, str]]:
        """Returns recent messages formatted for Ollama chat API."""
        rows = self.db.fetch_all(
            "SELECT role, content FROM campaign_messages WHERE campaign_id = ? AND is_active = 1 ORDER BY created_at DESC LIMIT ?",
            (campaign_id, limit)
        )
        # Reverse so oldest first (chronological for LLM context)
        return [{"role": r["role"], "content": r["content"]} for r in reversed(rows)]

    def get_pages(self, campaign_id: str) -> List[Dict[str, Any]]:
        """Aggregates player and DM turns into structured pages for Infinite Book layout."""
        rows = self.db.fetch_all(
            "SELECT role, content, page_number, coordinates_json, choices_json, mechanics, created_at FROM campaign_messages WHERE campaign_id = ? AND is_active = 1 ORDER BY page_number ASC, created_at ASC",
            (campaign_id,)
        )
        
        pages = {}
        for r in rows:
            p_num = r["page_number"]
            if p_num not in pages:
                pages[p_num] = {
                    "page_number": p_num,
                    "player_text": "",
                    "dm_text": "",
                    "coordinates": json.loads(r["coordinates_json"]) if r["coordinates_json"] else {"x": 2, "y": 4},
                    "choices": json.loads(r["choices_json"]) if r["choices_json"] else [],
                    "mechanics": r["mechanics"] or "",
                    "created_at": r["created_at"]
                }
            if r["role"] == "user":
                clean_text = re.sub(r'\[MECÁNICA:.*?\]', '', r["content"]).strip()
                pages[p_num]["player_text"] = clean_text
            else:
                pages[p_num]["dm_text"] = r["content"]
                if r["mechanics"]:
                    pages[p_num]["mechanics"] = r["mechanics"]
                if r["choices_json"]:
                    pages[p_num]["choices"] = json.loads(r["choices_json"])
                if r["coordinates_json"]:
                    pages[p_num]["coordinates"] = json.loads(r["coordinates_json"])
                    
        return sorted(list(pages.values()), key=lambda x: x["page_number"])

    def rollback_to_page(self, campaign_id: str, page_number: int) -> int:
        """Deactivates all pages after the specified page number, effectively branching/rolling back."""
        query = "UPDATE campaign_messages SET is_active = 0 WHERE campaign_id = ? AND page_number > ?"
        return self.db.execute(query, (campaign_id, page_number))


class MemoryVectorRepository:
    def __init__(self, db_manager: DatabaseManager):
        self.db = db_manager

    def add_vector(self, campaign_id: str, content: str, embedding: List[float], source_event_id: Optional[str] = None) -> None:
        import uuid
        vec_id = str(uuid.uuid4())
        self.db.execute(
            """INSERT INTO memory_vectors (id, campaign_id, content, embedding_json, source_event_id)
               VALUES (?, ?, ?, ?, ?)""",
            (vec_id, campaign_id, content, json.dumps(embedding), source_event_id)
        )

    def search_similar(self, campaign_id: str, query_vector: List[float], top_k: int = 3) -> List[Dict[str, Any]]:
        if not query_vector:
            return []
            
        def cosine_similarity(v1, v2):
            import math
            dot = sum(a*b for a, b in zip(v1, v2))
            norm1 = math.sqrt(sum(a*a for a in v1))
            norm2 = math.sqrt(sum(a*a for a in v2))
            if norm1 == 0 or norm2 == 0:
                return 0.0
            return dot / (norm1 * norm2)

        rows = self.db.fetch_all(
            "SELECT content, embedding_json FROM memory_vectors WHERE campaign_id = ?",
            (campaign_id,)
        )
        
        scored_results = []
        for r in rows:
            try:
                emb = json.loads(r["embedding_json"])
                sim = cosine_similarity(query_vector, emb)
                scored_results.append({"content": r["content"], "similarity": sim})
            except Exception:
                continue
                
        scored_results.sort(key=lambda x: x["similarity"], reverse=True)
        return scored_results[:top_k]

File: ai-dungeon-master-offline/app/frontend/src-tauri/src/repositories.rs
use std::collections::HashMap;
use rusqlite::{params, Connection, Result, Row};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;
use regex::Regex;
use rayon::prelude::*;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Campaign {
    pub id: String,
    pub name: String,
    pub system: String,
    pub tone: String,
    pub narrative_summary: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Character {
    pub id: String,
    pub campaign_id: Option<String>,
    pub name: String,
    pub class: String,
    pub race: String,
    pub background: String,
    pub level: i32,
    pub hp_current: i32,
    pub hp_max: i32,
    pub armor_class: i32,
    pub stats: Value,
    pub inventory: Value,
    pub xp: i32,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct CampaignPage {
    pub page_number: i32,
    pub player_text: String,
    pub dm_text: String,
    pub coordinates: Value,
    pub choices: Value,
    pub mechanics: String,
    pub created_at: String,
}

pub struct CampaignRepository;

impl CampaignRepository {
    pub fn create(conn: &Connection, name: &str, system: &str, tone: &str) -> Result<Campaign> {
        let id = Uuid::new_v4().to_string();
        conn.execute(
            "INSERT INTO campaigns (id, name, system, tone, narrative_summary) VALUES (?, ?, ?, ?, '')",
            params![id, name, system, tone],
        )?;
        Self::get_by_id(conn, &id)
    }

    pub fn get_by_id(conn: &Connection, id: &str) -> Result<Campaign> {
        conn.query_row(
            "SELECT id, name, system, tone, narrative_summary, created_at, updated_at FROM campaigns WHERE id = ?",
            params![id],
            |row| {
                Ok(Campaign {
                    id: row.get(0)?,
                    name: row.get(1)?,
                    system: row.get(2)?,
                    tone: row.get(3)?,
                    narrative_summary: row.get(4)?,
                    created_at: row.get(5)?,
                    updated_at: row.get(6)?,
                })
            },
        )
    }

    pub fn list_all(conn: &Connection) -> Result<Vec<Campaign>> {
        let mut stmt = conn.prepare("SELECT id, name, system, tone, narrative_summary, created_at, updated_at FROM campaigns ORDER BY created_at DESC")?;
        let rows = stmt.query_map([], |row| {
            Ok(Campaign {
                id: row.get(0)?,
                name: row.get(1)?,
                system: row.get(2)?,
                tone: row.get(3)?,
                narrative_summary: row.get(4)?,
                created_at: row.get(5)?,
                updated_at: row.get(6)?,
            })
        })?;
        let mut campaigns = Vec::new();
        for c in rows {
            campaigns.push(c?);
        }
        Ok(campaigns)
    }

    pub fn update_summary(conn: &Connection, id: &str, summary: &str) -> Result<()> {
        conn.execute(
            "UPDATE campaigns SET narrative_summary = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?",
            params![summary, id],
        )?;
        Ok(())
    }
}

pub struct CharacterRepository;

impl CharacterRepository {
    fn map_row(row: &Row) -> Result<Character> {
        let stats_str: String = row.get(10)?;
        let stats: Value = serde_json::from_str(&stats_str).unwrap_or(Value::Null);
        let inv_str: String = row.get(11)?;
        let inventory: Value = serde_json::from_str(&inv_str).unwrap_or(Value::Null);

        Ok(Character {
            id: row.get(0)?,
            campaign_id: row.get(1)?,
            name: row.get(2)?,
            class: row.get(3)?,
            race: row.get(4)?,
            background: row.get(5)?,
            level: row.get(6)?,
            hp_current: row.get(7)?,
            hp_max: row.get(8)?,
            armor_class: row.get(9)?,
            stats,
            inventory,
            xp: row.get(12)?,
        })
    }

    pub fn create(
        conn: &Connection,
        campaign_id: Option<&str>,
        name: &str,
        char_class: &str,
        race: &str,
        background: &str,
        hp_max: i32,
        armor_class: i32,
        stats: &Value,
        inventory: &Value,
    ) -> Result<Character> {
        let id = Uuid::new_v4().to_string();
        let stats_str = stats.to_string();
        let inv_str = inventory.to_string();

        conn.execute(
            "INSERT INTO characters (id, campaign_id, name, class, race, background, level, hp_current, hp_max, armor_class, stats_json, inventory_json, xp)
             VALUES (?, ?, ?, ?, ?, ?, 1, ?, ?, ?, ?, ?, 0)",
            params![id, campaign_id, name, char_class, race, background, hp_max, hp_max, armor_class, stats_str, inv_str],
        )?;

        Self::get_by_id(conn, &id)
    }

    pub fn get_by_id(conn: &Connection, id: &str) -> Result<Character> {
        conn.query_row(
            "SELECT id, campaign_id, name, class, race, background, level, hp_current, hp_max, armor_class, stats_json, inventory_json, xp FROM characters WHERE id = ?",
            params![id],
            Self::map_row,
        )
    }

    pub fn get_by_campaign(conn: &Connection, campaign_id: &str) -> Result<Vec<Character>> {
        let mut stmt = conn.prepare(
            "SELECT id, campaign_id, name, class, race, background, level, hp_current, hp_max, armor_class, stats_json, inventory_json, xp FROM characters WHERE campaign_id = ?"
        )?;
        let rows = stmt.query_map(params![campaign_id], Self::map_row)?;
        let mut chars = Vec::new();
        for r in rows {
            chars.push(r?);
        }
        Ok(chars)
    }

    pub fn get_all(conn: &Connection) -> Result<Vec<Character>> {
        let mut stmt = conn.prepare(
            "SELECT id, campaign_id, name, class, race, background, level, hp_current, hp_max, armor_class, stats_json, inventory_json, xp FROM characters"
        )?;
        let rows = stmt.query_map([], Self::map_row)?;
        let mut chars = Vec::new();
        for r in rows {
            chars.push(r?);
        }
        Ok(chars)
    }

    pub fn get_global_characters(conn: &Connection) -> Result<Vec<Character>> {
        let mut stmt = conn.prepare(
            "SELECT id, campaign_id, name, class, race, background, level, hp_current, hp_max, armor_class, stats_json, inventory_json, xp FROM characters WHERE campaign_id IS NULL"
        )?;
        let rows = stmt.query_map([], Self::map_row)?;
        let mut chars = Vec::new();
        for r in rows {
            chars.push(r?);
        }
        Ok(chars)
    }

    pub fn assign_to_campaign(conn: &Connection, char_id: &str, campaign_id: &str) -> Result<bool> {
        let count = conn.execute(
            "UPDATE characters SET campaign_id = ? WHERE id = ?",
            params![campaign_id, char_id],
        )?;
        Ok(count > 0)
    }

    pub fn update_hp(conn: &Connection, char_id: &str, hp: i32) -> Result<bool> {
        let count = conn.execute(
            "UPDATE characters SET hp_current = ? WHERE id = ?",
            params![hp, char_id],
        )?;
        Ok(count > 0)
    }

    pub fn update_xp(conn: &Connection, char_id: &str, xp_gain: i32) -> Result<Value> {
        let char_data = Self::get_by_id(conn, char_id)?;
        let xp_thresholds = [
            0, 300, 900, 2700, 6500, 14000, 23000, 34000, 48000, 64000, 85000, 100000,
            120000, 140000, 165000, 195000, 225000, 265000, 305000, 355000,
        ];
        let current_xp = char_data.xp + xp_gain;
        let current_level = char_data.level;
        let mut new_level = current_level;
        for (i, &threshold) in xp_thresholds.iter().enumerate() {
            if current_xp >= threshold {
                new_level = (i + 1) as i32;
            }
        }
        conn.execute(
            "UPDATE characters SET xp = ?, level = ? WHERE id = ?",
            params![current_xp, new_level, char_id],
        )?;

        Ok(serde_json::json!({
            "xp": current_xp,
            "level": new_level,
            "leveled_up": new_level > current_level
        }))
    }
}

pub struct DiceRollRepository;

impl DiceRollRepository {
    pub fn save_roll(
        conn: &Connection,
        campaign_id: &str,
        actor_id: &str,
        actor_name: &str,
        formula: &str,
        result: &Value,
    ) -> Result<Value> {
        let roll_id = Uuid::new_v4().to_string();
        let result_str = result.to_string();
        conn.execute(
            "INSERT INTO dice_rolls (id, campaign_id, actor_id, actor_name, formula, result_json) VALUES (?, ?, ?, ?, ?, ?)",
            params![roll_id, campaign_id, actor_id, actor_name, formula, result_str],
        )?;
        Ok(serde_json::json!({
            "id": roll_id,
            "campaign_id": campaign_id,
            "actor_id": actor_id,
            "actor_name": actor_name,
            "formula": formula,
            "result": result
        }))
    }
}

pub struct MessageRepository;

impl MessageRepository {
    pub fn add_message(
        conn: &Connection,
        campaign_id: &str,
        role: &str,
        content: &str,
        character_id: Option<&str>,
        page_number: i32,
        coordinates_json: Option<&str>,
        choices_json: Option<&str>,
        mechanics: Option<&str>,
    ) -> Result<()> {
        let msg_id = Uuid::new_v4().to_string();
        conn.execute(
            "INSERT INTO campaign_messages (id, campaign_id, character_id, role, content, page_number, coordinates_json, choices_json, mechanics, is_active)
             VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 1)",
            params![
                msg_id,
                campaign_id,
                character_id,
                role,
                content,
                page_number,
                coordinates_json,
                choices_json,
                mechanics
            ],
        )?;
        Ok(())
    }

    pub fn get_recent_history(conn: &Connection, campaign_id: &str, limit: i32) -> Result<Vec<Value>> {
        let mut stmt = conn.prepare(
            "SELECT role, content FROM campaign_messages WHERE campaign_id = ? AND is_active = 1 ORDER BY created_at DESC LIMIT ?"
        )?;
        let rows = stmt.query_map(params![campaign_id, limit], |row| {
            let role: String = row.get(0)?;
            let content: String = row.get(1)?;
            Ok(serde_json::json!({
                "role": role,
                "content": content
            }))
        })?;
        let mut history = Vec::new();
        for r in rows {
            history.push(r?);
        }
        history.reverse();
        Ok(history)
    }

    pub fn get_pages(conn: &Connection, campaign_id: &str) -> Result<Vec<CampaignPage>> {
        let mut stmt = conn.prepare(
            "SELECT role, content, page_number, coordinates_json, choices_json, mechanics, created_at
             FROM campaign_messages WHERE campaign_id = ? AND is_active = 1 ORDER BY page_number ASC, created_at ASC"
        )?;

        let rows = stmt.query_map(params![campaign_id], |row| {
            let role: String = row.get(0)?;
            let content: String = row.get(1)?;
            let page_number: i32 = row.get(2)?;
            let coords_str: Option<String> = row.get(3)?;
            let choices_str: Option<String> = row.get(4)?;
            let mechanics: Option<String> = row.get(5)?;
            let created_at: String = row.get(6)?;

            Ok((role, content, page_number, coords_str, choices_str, mechanics, created_at))
        })?;

        let mut pages_map: HashMap<i32, CampaignPage> = HashMap::new();
        let re_mech = Regex::new(r"\[MECÁNICA:.*?\]").unwrap();

        for r_res in rows {
            let (role, content, page_number, coords_str, choices_str, mechanics_opt, created_at) = r_res?;
            let coords = coords_str
                .and_then(|s| serde_json::from_str(&s).ok())
                .unwrap_or_else(|| serde_json::json!({"x": 2, "y": 4}));
            let choices = choices_str
                .and_then(|s| serde_json::from_str(&s).ok())
                .unwrap_or_else(|| serde_json::json!([]));
            let mechanics = mechanics_opt.unwrap_or_default();

            let entry = pages_map.entry(page_number).or_insert_with(|| CampaignPage {
                page_number,
                player_text: String::new(),
                dm_text: String::new(),
                coordinates: coords.clone(),
                choices: choices.clone(),
                mechanics: mechanics.clone(),
                created_at: created_at.clone(),
            });

            if role == "user" {
                let clean_text = re_mech.replace_all(&content, "").to_string();
                entry.player_text = clean_text.trim().to_string();
            } else {
                entry.dm_text = content;
                if !mechanics.is_empty() {
                    entry.mechanics = mechanics;
                }
                if choices != serde_json::json!([]) {
                    entry.choices = choices;
                }
                if coords != serde_json::json!({"x": 2, "y": 4}) {
                    entry.coordinates = coords;
                }
            }
        }

        let mut pages: Vec<CampaignPage> = pages_map.into_values().collect();
        pages.sort_by_key(|p| p.page_number);
        Ok(pages)
    }

    pub fn rollback_to_page(conn: &Connection, campaign_id: &str, page_number: i32) -> Result<usize> {
        let count = conn.execute(
            "UPDATE campaign_messages SET is_active = 0 WHERE campaign_id = ? AND page_number > ?",
            params![campaign_id, page_number],
        )?;
        Ok(count)
    }
}

pub struct MemoryVectorRepository;

impl MemoryVectorRepository {
    pub fn add_vector(
        conn: &Connection,
        campaign_id: &str,
        content: &str,
        embedding: &[f64],
        source_event_id: Option<&str>,
    ) -> Result<()> {
        let vec_id = Uuid::new_v4().to_string();
        let embedding_val = serde_json::json!(embedding);
        let emb_str = embedding_val.to_string();
        conn.execute(
            "INSERT INTO memory_vectors (id, campaign_id, content, embedding_json, source_event_id) VALUES (?, ?, ?, ?, ?)",
            params![vec_id, campaign_id, content, emb_str, source_event_id],
        )?;
        Ok(())
    }

    pub fn search_similar(
        conn: &Connection,
        campaign_id: &str,
        query_vector: &[f64],
        top_k: usize,
    ) -> Result<Vec<Value>> {
        if query_vector.is_empty() {
            return Ok(Vec::new());
        }

        let mut stmt = conn.prepare(
            "SELECT content, embedding_json FROM memory_vectors WHERE campaign_id = ?"
        )?;
        let rows = stmt.query_map(params![campaign_id], |row| {
            let content: String = row.get(0)?;
            let emb_str: String = row.get(1)?;
            Ok((content, emb_str))
        })?;

        let rows_data: Vec<(String, String)> = rows.filter_map(|r| r.ok()).collect();

        fn cosine_similarity(v1: &[f64], v2: &[f64]) -> f64 {
            let dot: f64 = v1.iter().zip(v2.iter()).map(|(a, b)| a * b).sum();
            let norm1: f64 = v1.iter().map(|x| x * x).sum::<f64>().sqrt();
            let norm2: f64 = v2.iter().map(|x| x * x).sum::<f64>().sqrt();
            if norm1 == 0.0 || norm2 == 0.0 {
                return 0.0;
            }
            dot / (norm1 * norm2)
        }

        let mut scored_results: Vec<Value> = rows_data
            .into_par_iter()
            .filter_map(|(content, emb_str)| {
                if let Ok(Value::Array(arr)) = serde_json::from_str::<Value>(&emb_str) {
                    let db_vector: Vec<f64> = arr.iter().filter_map(|v| v.as_f64()).collect();
                    if db_vector.len() == query_vector.len() {
                        let score = cosine_similarity(query_vector, &db_vector);
                        return Some(serde_json::json!({
                            "content": content,
                            "similarity": score
                        }));
                    }
                }
                None
            })
            .collect();

        // Sort by similarity descending
        scored_results.sort_by(|a, b| {
            let sim_a = a["similarity"].as_f64().unwrap_or(0.0);
            let sim_b = b["similarity"].as_f64().unwrap_or(0.0);
            sim_b.partial_cmp(&sim_a).unwrap_or(std::cmp::Ordering::Equal)
        });

        if scored_results.len() > top_k {
            scored_results.truncate(top_k);
        }

        Ok(scored_results)
    }
}
